pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java

```diff
@@ -0,0 +1,298 @@
+      BiMap<String, String> serverToEndpoints, String tableNameWithType, Set<String> completedSegments)
+      throws URISyntaxException {
+    Set<String> remainingSegments = new HashSet<>(completedSegments);
+    List<String> urls = new ArrayList<>();
+    for (Map.Entry<String, List<String>> entry : serverToSegments.entrySet()) {
+      if (remainingSegments.isEmpty()) {
+        break;
+      }
+      String server = entry.getKey();
+      List<String> segmentNames = entry.getValue();
+      URIBuilder uriBuilder = new URIBuilder(serverToEndpoints.get(server)).setPath(
+          String.format("/tables/%s/validDocIdMetadata", tableNameWithType));
+      int completedSegmentCountPerServer = 0;
+      for (String segmentName : segmentNames) {
+        if (remainingSegments.contains(segmentName)) {
+          completedSegmentCountPerServer++;
+          uriBuilder.addParameter("segmentNames", segmentName);
+          remainingSegments.remove(segmentName);
+        }
+      }
+      if (completedSegmentCountPerServer > 0) {
+        // only add to the list if the server has completed segments
+        urls.add(uriBuilder.toString());
+      }
+    }
+    return urls;
+  }
+
+  private List<SegmentZKMetadata> getCompletedSegments(String tableNameWithType,
+      Map<String, String> compactionConfigs) {
+    List<SegmentZKMetadata> completedSegments = new ArrayList<>();
+    String bufferPeriod =
+        compactionConfigs.getOrDefault(UpsertCompactionTask.BUFFER_TIME_PERIOD_KEY, DEFAULT_BUFFER_PERIOD);
+    long bufferMs = TimeUtils.convertPeriodToMillis(bufferPeriod);
+    List<SegmentZKMetadata> allSegments = _clusterInfoAccessor.getSegmentsZKMetadata(tableNameWithType);
+    for (SegmentZKMetadata segment : allSegments) {
+      CommonConstants.Segment.Realtime.Status status = segment.getStatus();
+      // initial segments selection based on status and age
+      if (status.isCompleted() && (segment.getEndTimeMs() <= (System.currentTimeMillis() - bufferMs))) {
+        completedSegments.add(segment);
+      }
+    }
+    return completedSegments;
+  }
+
+  @VisibleForTesting
+  public static int getMaxTasks(String taskType, String tableNameWithType, Map<String, String> taskConfigs) {
+    int maxTasks = Integer.MAX_VALUE;
+    String tableMaxNumTasksConfig = taskConfigs.get(MinionConstants.TABLE_MAX_NUM_TASKS_KEY);
+    if (tableMaxNumTasksConfig != null) {
+      try {
+        maxTasks = Integer.parseInt(tableMaxNumTasksConfig);
+      } catch (Exception e) {
+        LOGGER.warn("MaxNumTasks have been wrongly set for table : {}, and task {}", tableNameWithType, taskType);
+      }
+    }
+    return maxTasks;
+  }
+
+  private static final String[] VALID_CONFIG_KEYS = {
+      UpsertCompactionTask.BUFFER_TIME_PERIOD_KEY, UpsertCompactionTask.INVALID_RECORDS_THRESHOLD_PERCENT,
+      UpsertCompactionTask.INVALID_RECORDS_THRESHOLD_COUNT
+  };
+
+  private Map<String, String> getCompactionConfigs(Map<String, String> taskConfig) {
```

Comment:
(minor) There is no need to extract this map. We can directly read the `taskConfig`